# Spatial transcriptomics

```{toctree}
:maxdepth: 1

notebooks/spatial/resolVI_tutorial
notebooks/spatial/DestVI_tutorial
notebooks/spatial/gimvi_tutorial
notebooks/spatial/tangram_scvi_tools
notebooks/spatial/stereoscope_heart_LV_tutorial
notebooks/spatial/cell2location_lymph_node_spatial_tutorial
```

```{customcard}
:path: notebooks/spatial/resolVI_tutorial
:tags: Analysis, Integration, Transfer-learning, Dimensionality-reduction

TODO: add description
```

```{customcard}
:path: notebooks/spatial/DestVI_tutorial
:tags: Deconvolution, Modality-imputation, Differential-comparison

Perform multi-resolution analysis on spatial transcriptomics data with DestVI
```

```{customcard}
:path: notebooks/spatial/gimvi_tutorial
:tags: Modality-imputation

Use gimVI to impute missing genes in spatial data
```

```{customcard}
:path: notebooks/spatial/tangram_scvi_tools
:tags: Deconvolution

Use Tangram to map spatial transcriptomics data
```

```{customcard}
:path: notebooks/spatial/stereoscope_heart_LV_tutorial
:tags: Deconvolution

Go through the Stereoscope workflow to map single-cell data
```

```{customcard}
:path: notebooks/spatial/cell2location_lymph_node_spatial_tutorial
:tags:

Spatially map lymph node cell types using Cell2location
```
