References
----------

.. [Ashuach22] Tal Ashauch, Daniel A Reidenbach, Adam Gayoso, Nir Yosef (2022),
   *PeakVI: A deep generative model for single-cell chromatin accessibility analysis*,
   `Cell Reports Methods <https://doi.org/10.1016/j.crmeth.2022.100182>`__.

.. [Gayoso22] Adam Gayoso*, Romain Lopez*, Galen Xing*, Pierre Boyeau, Valeh Valiollah Pour Amiri, Justin Hong, Katherine Wu, Michael Jayasuriya, Edouard Mehlman, Maxime Langevin, Yining Liu, Jules Samaran, Gabriel Misrachi, Achille Nazaret, Oscar Clivio, Chenling Xu, Tal Ashuach, Mohammad Lotfollahi, Valentine Svensson, Eduardo da Veiga Beltrame, Vitalii Kleshchevnikov, Carlos Talavera-Lopez, Lior Pachter, Fabian J Theis, Aaron Streets, Michael I Jordan, Jeffrey Regier, and Nir Yosef (2022),
   *A Python library for probabilistic analysis of single-cell omics data*,
   `Nature Biotechnology <http://dx.doi.org/10.1038/s41587-021-01206-w>`__.

.. [Lotfollahi21] Mohammad Lotfollahi, Mohsen Naghipourfar, Malte D. Luecken, Matin Khajavi, Maren Büttner, Marco Wagenstetter, Ziga Avsec, Adam Gayoso, Nir Yosef, Marta Interlandi, Sergei Rybakov, Alexander V. Misharin, and Fabian J. Theis (2021),
   *Mapping single-cell data to reference atlases by transfer learning*,
   `Nature Biotechnology <https://www.nature.com/articles/s41587-021-01001-7>`__.

.. [Lopez21] Romain Lopez*, Baoguo Li*, Hadas Keren-Shaul*, Pierre Boyeau, Merav Kedmi, David Pilzer, Adam Jelinski, Eyal David, Allon Wagner, Yoseph Addadi, Michael I Jordan, Ido Amit, Nir Yosef. (2021),
   *Multi-resolution deconvolution of spatial transcriptomics data reveals continuous patterns of inflammation*,
   `biorxiv <https://www.biorxiv.org/content/10.1101/2021.04.28.441833v1>`__.

.. [AshuachGabitto21] Tal Ashauch*, Mariano I Gabitto*, Michael I Jordan, Nir Yosef (2021),
   *MultiVI: deep generative model for the integration of multi-modal data*,
   `biorxiv <https://www.biorxiv.org/content/10.1101/2021.08.20.457057v1>`__.

.. [GayosoSteier21] Adam Gayoso*, Zoë Steier*, Romain Lopez, Jeffrey Regier, Kristopher L Nazor, Aaron Streets, Nir Yosef (2021),
   *Joint probabilistic modeling of single-cell multi-omic data with totalVI*,
   `Nature Methods <https://www.nature.com/articles/s41592-020-01050-x>`__.

.. [Xu21] Chenling Xu*, Romain Lopez*, Edouard Mehlman*, Jeffrey Regier, Michael I. Jordan, Nir Yosef (2021),
   *Probabilistic harmonization and annotation of single-cell transcriptomics data with deep generative models*,
   `Molecular Systems Biology <https://www.embopress.org/doi/full/10.15252/msb.20209620>`__.

.. [Andersson20] Alma Andersson, Joseph Bergenstråhle, Michaela Asp, Ludvig Bergenstråhle, Aleksandra Jurek, José Fernández Navarro & Joakim Lundeberg (2020),
   *Single-cell and spatial transcriptomics enables probabilistic inference of cell type topography*,
   `Communications Biology <https://www.nature.com/articles/s42003-020-01247-y>`__.

.. [Svensson20] Valentine Svensson, Adam Gayoso, Nir Yosef, Lior Pachter (2020),
   *Interpretable factor models of single-cell RNA-seq via variational autoencoders*,
   `Bioinformatics <https://academic.oup.com/bioinformatics/article/36/11/3418/5807606>`__.

.. [Boyeau19] Pierre Boyeau, Romain Lopez, Jeffrey Regier, Adam Gayoso, Michael I. Jordan, Nir Yosef (2019),
   *Deep generative models for detecting differential expression in single cells*,
   `Machine Learning in Computational Biology (MLCB) <https://www.biorxiv.org/content/biorxiv/early/2019/10/04/794289.full.pdf>`__.

.. [Clivio19] Oscar Clivio, Romain Lopez, Jeffrey Regier, Adam Gayoso, Michael I. Jordan, Nir Yosef (2019),
   *Detecting zero-inflated genes in single-cell transcriptomics data*,
   `Machine Learning in Computational Biology (MLCB) <https://www.biorxiv.org/content/biorxiv/early/2019/10/10/794875.full.pdf>`__.

.. [Lopez19] Romain Lopez*, Achille Nazaret*, Maxime Langevin*, Jules Samaran*, Jeffrey Regier*, Michael I. Jordan, Nir Yosef (2019),
   *A joint model of unpaired data from scRNA-seq and spatial transcriptomics for imputing missing gene expression measurements*,
   `ICML Workshop on Computational Biology <https://arxiv.org/pdf/1905.02269.pdf>`__.

.. [Bernstein19] Nicholas J. Bernstein, , Nicole L. Fong, Irene Lam, Margaret A. Roy, David G. Hendrickson, and David R. Kelley (2020),
    *Solo: doublet identification in single-cell RNA-Seq via semi-supervised deep learning*,
    `Cell Systems <https://www.sciencedirect.com/science/article/pii/S2405471220301952>`__.

.. [Zhang19] Allen W. Zhang, Ciara O’Flanagan, Elizabeth A. Chavez, Jamie LP Lim, Nicholas Ceglia, Andrew McPherson, Matt Wiens et al. (2019),
   *Probabilistic cell-type assignment of single-cell RNA-seq for tumor microenvironment profiling*,
   `Nature Methods <https://www.nature.com/articles/s41592-019-0529-1?elqTrackId=12c8cef68e0741ef8422778b61>`__.

.. [Lopez18] Romain Lopez, Jeffrey Regier, Michael Cole, Michael I. Jordan, Nir Yosef (2018),
   *Deep generative modeling for single-cell transcriptomics*,
   `Nature Methods <https://www.nature.com/articles/s41592-018-0229-2.epdf?author_access_token=5sMbnZl1iBFitATlpKkddtRgN0jAjWel9jnR3ZoTv0P1-tTjoP-mBfrGiMqpQx63aBtxToJssRfpqQ482otMbBw2GIGGeinWV4cULBLPg4L4DpCg92dEtoMaB1crCRDG7DgtNrM_1j17VfvHfoy1cQ%3D%3D>`__.

.. [Blei03] David M. Blei, Andrew Y. Ng, Michael I. Jordan (2003),
   *Latent Dirichlet Allocation*,
   `Journal of Machine Learning Research <https://www.jmlr.org/papers/volume3/blei03a/blei03a.pdf>`__.
